### ISSUE TITLE
Incorrect error message on invalid union type assignation

### ISSUE THREAD

**Issue by lovasoa** [microsoft/TypeScript#55465]
### 🔎 Search Terms

error message union type assignation

### 🕗 Version & Regression Information

- This is the behavior in every version I tried

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=5.2.1-rc#code/MYewdgzgLgBMBcMDeAPGiDkBDDMA+MGARrvocBgL4wC8yK82VMAUKJLEYqpjtQT0IlqdYAG4gA

### 💻 Code

```ts
const c: {x : 'a' | 'b'  | 'c'} = {x:'a'} 
const b: {x: 'a'} | {x: 'b'} = c;
```

### 🙁 Actual behavior

The error message is

```
Type '{ x: "a" | "b" | "c"; }' is not assignable to type '{ x: "a"; } | { x: "b"; }'.
  Type '{ x: "a" | "b" | "c"; }' is not assignable to type '{ x: "b"; }'.
    Types of property 'x' are incompatible.
      Type '"a" | "b" | "c"' is not assignable to type '"b"'.
        Type '"a"' is not assignable to type '"b"'.(2322)
```

In this simple case, it's easy to see where the actual problem is, but with large unions, it becomes hard to know where the error comes from.

### 🙂 Expected behavior

The actual problem is that **type 'c' is not assignable to 'a' | 'b'**, and the error message should mention it. Maybe the full trace should look like this:

```
Type '{ x: "a" | "b" | "c"; }' is not assignable to type '{ x: "a"; } | { x: "b"; }'.
    Types of property 'x' are incompatible.
      Type '"a" | "b" | "c"' is not assignable to type '"a" | "b"'.
        Type '"c"' is not assignable to type '"a" | "b"'.(2322)
```

### Additional information about the issue

When the unions are large and aliased, the misleading message makes errors hard to debug.

**Comment by lovasoa** [comment]
Maybe it's a bit similar, yes. But there isn't any non-distinct discriminating property here.

**Comment by lovasoa** [comment]
@MartinJohns 

> But this would imply that a value typed `"a" | "b"` can be used, which is not the case.

I'm not sure I understand. A value of type `"a" | "b"` **can** be used for property `x`. The following would be correct, for instance :

```ts
const c: {x : 'a' | 'b'} = {x:'a'} 
const b: {x: 'a'} | {x: 'b'} = c;
```

**Comment by fatcerberus** [comment]
In general, `{ x: A } | { x: B }` is not equivalent to `{ x: A | B }`.  In some limited cases where TS can prove it's safe it will allow an assignment, but as a general rule those types aren't interchangeable and the compiler therefore won't merge them for the purpose of producing error messages.

**Comment by lovasoa** [comment]
To come back to the main topic of this issue, the incorrect error message seems to come from the way typescript checks assignation to union types. It checks each of the union members individually (which is correct), but if none match, it reports a mismatch between the source type and the **last** member of the union, which is incorrect and confusing.

**Comment by fatcerberus** [comment]
> if none match, it reports a mismatch between the source type and the last member of the union

IIRC it only does this for discriminated unions, which is what you have in the OP.

**Comment by lovasoa** [comment]
Yes, running tests, I can see that. I see that

```ts
type X<T> = {x: T};

function f<A, B>(x: X<A|B>): X<A> | X<B> {
    return x
}
```

does not typecheck, whereas it should.
